adv_op.c

```diff
@@ -15,13 +16,51 @@ void p_char(stack_t **stack, unsigned int line_number)
 	int ascii_value;
 
 	if (!stack || (!*stack))
-		dprintf(STDERR_FILENO, NO_PCHAR, line_number), exit(EXIT_FAILURE);
+	{
+		dprintf(STDERR_FILENO, NO_PCHAR, line_number);
+		free_stack(*stack);
+		EXIT_F
+		
+	}
 
 	ascii_value = (*stack)->n;
 
+	if (ascii_value >= 0 && ascii_value  <= 127)
+		printf("%c\n", ascii_value);
+	else
+	{
+		dprintf(STDERR_FILENO, NO_VAL, line_number);
+		EXIT_F
```

Comment:
You might be able to save 3 lines (if it works) by moving the EXIT_F onto the dprintf line with a comma, then removing the else brackets